rdtools/degradation.py

```diff
@@ -80,3 +80,293 @@ def degradation_with_ols(normalized_energy):
     }
 
     return degradation
+
+      
+###  NEW classical decomposition routine from Chris Deline
+def degradation_classical_decomposition(normalized_energy, interpolate_flag = False):
+    # (degradation) = degradation_classical_decomposition(normalized_energy, interpolate_flag = False)
+    #
+    # classical decomponsition method using statsmodel.api
+    # from Dirk Jordan (NREL) 12/10/16
+    # input parameters:  
+    #       dataframe['normalized_energy'] : corrected performance ratio.
+    #       dataframe.index : timestamp format, monthly frequency
+    #       interpolate_flag:  boolean flag to either interpolate missing data or fill with the median value (default)
+    #output parameters:
+    #   degradation : tuple of annual degradation rate and standard error
+    #              'Rd_pct', 'slope', 'intercept', 'Rd_stderr_pct', 'dataframe'  
+    #   dataframe:  dataframe with annual rolling mean values
+    
+    # TODO:  check timebase for other input time options (day, week)    
+
+    
+    ################################################################################################################################
+    # Classical decomposition approach 
+
+    #need to choose fill methods: interpolate, or replace NaN's with the median value.
+    #interpolate_flag = True
+    
+    
+    dataframe = pd.DataFrame()
+    dataframe['normalized_energy'] = normalized_energy
+    #try to extract month information from the dataframe index. Fill missing data with median value
+    try:
+        energy_median = normalized_energy.median()
+        dataframe = dataframe.resample('MS').mean()
+        
+        if interpolate_flag:
+            dataframe = dataframe.interpolate()    
+        else:  #append the median value to missing months
+            dataframe = dataframe.fillna(value=energy_median)
+        
+        dataframe['Month'] = np.arange(0, len(dataframe))
+
+    except:
+        # if you don't pass a datetime index, it assumes that there is data for each month
+        dataframe['Month'] = np.arange(0, len(dataframe))
+        
+    if pd.infer_freq(dataframe.index) == 'MS' and len(normalized_energy) > 12:
+        dataframe = dataframe.rolling(window = 12, center = True).mean()
+        
+    dataframe = dataframe.dropna()
+    y2=dataframe['normalized_energy']
+    x2 = dataframe['Month']
+    
+
+
+    if len(dataframe) ==0:
+        degradation_values = {}
+
+        print '\nNot enough data for seasonal decomposition:'
+        
+        
+    else:        
+        # OLS regression with constant
+        x2=sm.add_constant(x2)
+        model2 = sm.OLS(y2,x2).fit()
+    
+        b_cd = model2.params['const']
+        m_cd = model2.params['Month']
+        SE_b_cd = model2.bse['const']
+        SE_m_cd = model2.bse['Month']
+    
+        Rd_cd = (m_cd * 12) / b_cd * 100
+        SE_Rd_cd = np.sqrt(np.power(SE_m_cd * 12/b_cd, 2) + np.power((-12*m_cd/b_cd**2) * SE_b_cd, 2))*100
```

Comment:
This OLS uncertainty calculation is also in `degradation_with_ols` function. In order to not duplicate logic, this can be factored-out into another function, perhaps `ols_rd_uncertainty(ols_model)` -- which takes the ols model class as input.